src/thor/trace_route_action.cc

```diff
@@ -381,91 +397,140 @@ thor_worker_t::map_match(valhalla_request_t& request,
-      odin::Location origin;
-      PathLocation::toPBF(matcher->state_container()
-                              .state(first_result_with_state->stateid)
-                              .candidate(),
-                          &origin, *reader);
-      odin::Location destination;
-      PathLocation::toPBF(matcher->state_container()
-                              .state(last_result_with_state->stateid)
-                              .candidate(),
-                          &destination, *reader);
-
-      bool found_origin = false;
-      for (const auto& e : origin.path_edges()) {
-        if (e.graph_id() == path_edges.front().edgeid) {
-          found_origin = true;
-          break;
+    // trace_attributes always returns a single trip path and may have discontinuities
+    if (request.options.action() == odin::DirectionsOptions::trace_attributes) {
+      trip_path = path_map_match(match_results, controller, path_edges, route_discontinuities);
+      trip_paths.emplace_back(trip_path);
+      // trace_route can return multiple trip paths and cannot have discontinuities
+    } else {
+      auto origin = match_results.begin();
+      auto destination = match_results.begin();
+
+      size_t last_index = 0;
+      // for each edge in the path
+      for (size_t i = 0; i < path_edges.size(); ++i) {
+        const auto& path_edge = path_edges[i];
+        // while we still have trace points that are on this current edge of the path
+        while (destination != match_results.end() && path_edge.edgeid == destination->edgeid) {
+          // if it's not the first trace point, and it's a break, make a trip path leg
+          if (origin != destination &&
+              request.options.shape(destination - match_results.begin()).type() ==
+                  odin::Location::kBreak) {
+            std::vector<PathInfo> sub_path_edges(path_edges.begin() + last_index,
+                                                 path_edges.begin() + i + 1);
+            trip_path = thor::TripLegBuilder::Build(controller, matcher->graphreader(), mode_costing,
+                                                    sub_path_edges,
+                                                    *request.options.mutable_shape(
+                                                        origin - match_results.begin()),
+                                                    *request.options.mutable_shape(
+                                                        destination - match_results.begin()),
+                                                    std::list<odin::Location>{}, interrupt,
+                                                    &route_discontinuities);
+            trip_paths.emplace_back(trip_path);
+            // beginning of next leg will be the end of this leg
+            origin = destination;
+            // store the starting index of the path_edges
+            last_index = i;
+          }
+          // increment to next trace point
+          ++destination;
```

Comment:
L406 to L436 is the core of the change here. We loop over edges and `shape` locations, and generate a trippath whenever we encounter a `break`. Most of the rest of the PR is refactoring to enable returning multiple trippaths in a mapmatch result.